packages/parse5-htmlparser2-tree-adapter/lib/index.ts

```diff
@@ -0,0 +1,264 @@
+    const prevNode = parentNode.children[parentNode.children.indexOf(referenceNode) - 1];
+
+    if (prevNode && isText(prevNode)) {
+        prevNode.data += text;
+    } else {
+        insertBefore(parentNode, createTextNode(text), referenceNode);
+    }
+}
+
+export function adoptAttributes(recipient: Element, attrs: Attribute[]): void {
+    for (let i = 0; i < attrs.length; i++) {
+        const attrName = attrs[i].name;
+
+        if (typeof recipient.attribs[attrName] === 'undefined') {
+            recipient.attribs[attrName] = attrs[i].value;
+            recipient['x-attribsNamespace']![attrName] = attrs[i].namespace!;
+            recipient['x-attribsPrefix']![attrName] = attrs[i].prefix!;
+        }
+    }
+}
+
+//Tree traversing
+export function getFirstChild(node: NodeWithChildren): Node | null {
+    return node.children[0];
+}
+
+export function getChildNodes(node: NodeWithChildren): Node[] {
+    return node.children;
+}
+
+export function getParentNode(node: Node): NodeWithChildren | null {
+    return node.parent;
+}
+
+export function getAttrList(element: Element): Attribute[] {
+    return element.attributes;
+}
+
+//Node data
+export function getTagName(element: Element): string {
+    return element.name;
+}
+
+export function getNamespaceURI(element: Element): NS {
+    return element.namespace as NS;
+}
+
+export function getTextNodeContent(textNode: Text): string {
+    return textNode.data;
+}
+
+export function getCommentNodeContent(commentNode: Comment): string {
+    return commentNode.data;
+}
+
+export function getDocumentTypeNodeName(doctypeNode: ProcessingInstruction): string | null {
+    return doctypeNode['x-name'] ?? null;
+}
+
+export function getDocumentTypeNodePublicId(doctypeNode: ProcessingInstruction): string | null {
+    return doctypeNode['x-publicId'] ?? null;
+}
+
+export function getDocumentTypeNodeSystemId(doctypeNode: ProcessingInstruction): string | null {
+    return doctypeNode['x-systemId'] ?? null;
+}
+
+//Node types
+
+export function isDocumentTypeNode(node: Node): node is ProcessingInstruction {
+    return isDirective(node) && node.name === '!doctype';
+}
+
+// Source code location
+export function setNodeSourceCodeLocation(node: Node, location: ElementLocation | null): void {
+    if (location) {
+        node.startIndex = location.startOffset;
+        node.endIndex = location.endOffset;
+    }
+
+    node.sourceCodeLocation = location as any;
+}
+
+export function getNodeSourceCodeLocation(node: Node): ElementLocation | null | undefined {
+    return node.sourceCodeLocation as ElementLocation | null | undefined;
+}
+
+export function updateNodeSourceCodeLocation(node: Node, endLocation: Partial<ElementLocation>): void {
+    if (endLocation.endOffset != null) node.endIndex = endLocation.endOffset;
+
+    node.sourceCodeLocation = {
+        ...node.sourceCodeLocation,
+        ...endLocation,
+    } as any;
```

Comment:
what happens when you don't cast this? would be good to get it working with a strong type

another future nice to have too: no-explicit-any rule and drop all the usages 🙏

Reply:
I still have to add `startTag` and `endTag` in my types for `domhandler`.

Reply:
ah i get you. maybe you can at least cast it as the thing it wants (which may require a prior `as unknown`, i.e. `as unknown as ActualType`).

and possibly leave a TODO on it? so we don't forget

Reply:
Added TODO comments.